src/strategy.ts

```diff
@@ -0,0 +1,184 @@
+import { Strategy as OAuth2Strategy } from 'passport-oauth2';
+
+import { mapUserProfile } from './mapUserProfile';
+import { Profile, ProfileWithMetaData } from './models/profile';
+import { StrategyOption } from './models/strategyOptions';
+import { TwitterError } from './models/twitterError';
+import { TwitterUserInfoResponse } from './models/twitterUserInfo';
+
+export class Strategy extends OAuth2Strategy {
+  _userProfileURL: string;
+
+  /**
+   * Twitter strategy constructor
+   *
+   * The Twitter authentication strategy authenticates requests by delegating to
+   * Twitter using the OAuth 2.0 protocol.
+   *
+   * Applications must supply a `verify` callback which accepts an `accessToken`,
+   * `refreshToken` and service-specific `profile`, and then calls the `cb`
+   * callback supplying a `user`, which should be set to `false` if the
+   * credentials are not valid.  If an exception occured, `err` should be set.
+   *
+   * Required options:
+   *   - `clientID`      your Twitter application's App ID
+   *   - `clientSecret`  your Twitter application's App Secret
+   *   - `callbackURL`   URL to which Twitter will redirect the user after granting authorization
+   *   - `clientType`    your Twitter application (client) type, either `public` or `confidental`
+   *
+   * Examples:
+   *
+   *     passport.use(new TwitterStrategy({
+   *         clientType: 'confidential',
+   *         clientID: 'client-identification',
+   *         clientSecret: 'secret'
+   *         callbackURL: 'https://www.example.net/auth/twitter/callback'
+   *       },
+   *       function(accessToken, refreshToken, profile, cb) {
+   *         User.findOrCreate(..., function (err, user) {
+   *           cb(err, user);
+   *         });
+   *       }
+   *     ));
+   */
+  constructor(
+    options: StrategyOption,
+    verify: (
+      accessToken: string,
+      refreshToken: string,
+      profile: Profile,
+      done: (error: Error, user?: ProfileWithMetaData) => void
+    ) => void
+  ) {
+    options = options || {};
+    options.sessionKey = options.sessionKey || 'oauth:twitter';
+    const authorizationURL =
+      options.authorizationURL || 'https://twitter.com/i/oauth2/authorize';
+    const tokenURL =
+      options.tokenURL || 'https://api.twitter.com/2/oauth2/token';
+
+    let scope = options.scope || [];
```

Comment:
question: Isn't this a regression from #11 since you don't set `users.read` scope by default?
I am ok with handling this in a separate PR as long as it doesn't get released as stable; the original feature was untested and didn't consider `skipUserProfile` option.

Reply:
Yes it is regression. Thanks for the catch. I will handle it in separate PR as you propose.

Reply:
I just checked that version 1.1.0 has been released and tagged as `latest`.  I will fix it here in this PR.

Reply:
Also add some tests.

Reply:
@jnv Lessons learned while writing tests: 
- scopes passed via strategy constructor are not merged with scopes passed in authenticate function in passport-oauth2, scopes from authenticate function takes precedense.

I decided change this behaviour in Twitter strategy and to merge strategy constructor scope with authenticate function scope option.

Reply:
Hmm, I understand the thinking behind merging the scopes and I agree it seems like a better behavior, however my big concern is that it deviates from how all other strategies work, which will lead to unpleasant surprises. Not to mention it prevents you from removing scopes during authentication.

It's also an unnecessary breaking change - if I understand correctly, with default export fixed, we'd have a full backwards compatibility. 

I'd rather recommend to keep the default behavior or make scopes merging an opt-in behavior.

Reply:
I agree, I will keep the default behavior.